File: alarmmanagerdemo/src/main/java/com/meevii/alarmmanagerdemo/MainActivity.java
package com.meevii.alarmmanagerdemo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;

import com.meevii.alarmmanagerdemo.alarm.AlarmUtils;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.start).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlarmUtils.startAlarm(MainActivity.this, 100);
            }
        });

        findViewById(R.id.end).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                AlarmUtils.cancel(MainActivity.this);
            }
        });
    }
}

File: constraintlayout/src/main/java/com/wys/constraintlayout/MainActivity.kt
package com.wys.constraintlayout

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        runBlocking {
            val job=GlobalScope.launch {
                // launch new coroutine in background and continue
                delay(1000L) // non-blocking delay for 1 second (default time unit is ms)
                Log.d("MainActivity", "World") // print after delay
            }
            Log.d("MainActivity", "Hello") // main thread continues while coroutine is delayed
            job.join()
        }
//        runBlocking {
//            delay(2000L)
//        }
//        Thread.sleep(2000L) //
        Log.d("MainActivity", "block")

        image.setOnClickListener {
            imageHolder.setContentId(R.id.image)
        }

        replaceTextView.setOnClickListener {
            textView1.text = "abc"
        }
    }
}
